File: c++/sortings/merge_sort.cpp
#include <bits/stdc++.h>
using namespace std;
void merge(vector<int> *temparray,int low,int mid,int high)
{
	int array1_size=mid-low+1;
	int array2_size=high-mid;
	vector<int>array1(array1_size);
	vector<int>array2(array2_size);
	int actual_index=low;
	for(int i=0;i<=array1_size;i++)
	{
		array1[i]=(*temparray)[actual_index];
		actual_index++;
	}
	actual_index=mid+1;
	for(int i=0;i<=array2_size;i++)
	{
		array2[i]=(*temparray)[actual_index];
		actual_index++;
	}
	int a1=0,a2=0;
	actual_index=low;
	while(a1<array1_size && a2<array2_size)
	{
		if(array1[a1]<array2[a2])
		{
			(*temparray)[actual_index]=array1[a1];
			a1++;
		}
		else
		{
			(*temparray)[actual_index]=array2[a2];
			a2++;
		}
		actual_index++;
	}
	while(a1<array1_size)
	{
		(*temparray)[actual_index]=array1[a1];
		a1++;
		actual_index++;
	}
	while(a2<array2_size)
	{
		(*temparray)[actual_index]=array2[a2];
		a2++;
		actual_index++;
	}
}
void mergesort(vector<int> *temparray,int low,int high)
{
	if(low<high)
	{
		int mid=(low+high)/2;
		mergesort(temparray,low,mid);
		mergesort(temparray,mid+1,high);
		merge(temparray,low,mid,high);
	}
}

int main()
{
	vector <int> array;
	int size=0;
	cout<<"enter array size : ";
	cin>>size;
	cout<<"enter all elements : ";
	for(int i=0;i<size;i++)
	{
		int temp;
		cin>>temp;
		array.push_back(temp);
	}
	mergesort(&array,0,size-1);
	for(auto x:array)
	{
		cout<<x<<" ";
	}
	
	
}

File: python/sortings/merge_sort.py
def merge(array,left,right):
	mid=(left+right)//2
	list1=[array[i] for i in range(left,mid+1)]
	list2=[array[i] for i in range(mid+1,right+1)]
	l1=0
	l2=0
	i=left
	while(l1<len(list1) and l2<len(list2)):
		if(list1[l1]<list2[l2]):
			array[i]=list1[l1]
			l1+=1
		else:
			array[i]=list2[l2]
			l2+=1
		i+=1
	while(l1<len(list1)):
		array[i]=list1[l1]
		l1+=1
		i+=1
	while(l2<len(list2)):
		array[i]=list2[l2]
		l2+=1
		i+=1
def merge_sort(array,left,right):
	if(left<right):
		mid=(left+right)//2
		merge_sort(array,left,mid)
		merge_sort(array,mid+1,right)
		merge(array,left,right)
		
if __name__=="__main__":
	print("size of the array : ",end="")
	N=int(input())
	print("enter elements of the array : ",end="")
	array=list(map(int,input().split()))
	merge_sort(array,0,N-1)
	print(array)
	
